File: leetcode/c++/easy/_1752_check_if_array_is_sorted_and_rotated.cpp
// Leetcode Problem 1752

#include <gtest/gtest.h>

using namespace std;

class Solution {
public:
    bool check(vector<int>& nums) {
        bool rotationPoint = false;

        for (int i = 1; i < nums.size(); i++) {
            if ((nums[i-1] > nums[i] && rotationPoint)) {
                return false;
            } else if ((nums[i-1] > nums[i]) && !rotationPoint) {
                rotationPoint = true;
            }
        }

        if ((nums[0] < nums[nums.size() - 1]) && rotationPoint) {
            return false;
        }

        return true;
    }
};

class TestCheck : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestCheck, SimpleTestCase1) {
    vector<int> nums = {3, 4, 5, 1, 2};

    EXPECT_EQ(solution.check(nums), true);
}

TEST_F(TestCheck, SimpleTestCase2) {
    vector<int> nums = {2, 1, 3, 5};

    EXPECT_EQ(solution.check(nums), false);
}

TEST_F(TestCheck, SimpleTestCase3) {
    vector<int> nums = {1, 2, 3};

    EXPECT_EQ(solution.check(nums), true);
}

File: leetcode/rust/problems/src/easy/_1752_check_if_array_is_sorted_and_rotated.rs
// Problem 1752

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn check(nums: Vec<i32>) -> bool {
        let mut rotation_point = false;

        for i in 1..nums.len() {
            if nums[i-1] > nums[i] && rotation_point {
                return false;
            } else if nums[i-1] > nums[i] && !rotation_point {
                rotation_point = true;
            }
        }

        if rotation_point && nums[0] < nums[nums.len() - 1] {
            return false;
        }

        true
    }
}

#[test]
pub fn test_check() {
    let nums1 = vec![3, 4, 5, 1, 2];
    let nums2 = vec![2, 1, 3, 4];
    let nums3 = vec![1, 2, 3];

    assert_eq!(Solution::check(nums1), true);
    assert_eq!(Solution::check(nums2), false);
    assert_eq!(Solution::check(nums3), true);
}